### ISSUE TITLE
feature request: GetOrSet (w/ sync.Map.LoadOrStore semantics)

### ISSUE THREAD

**Issue by jdef** [allegro/bigcache#215]
Use case:
(1) goroutine1 checks for key K in the cache; it doesn't exist, dispatches a request w/ "cache-miss" flag
(2) goroutine2 (dispatcher) sees "cache-miss", computes needed value, saves it in cache

Using Get for (1) and Set for (2) there's a potential race wherein (1) could check for the same key K multiple times before (2) runs. This is more interesting when computing needed values result in resource accounting/allocation from a third party component (read: allocated "value" must be released when they are no longer needed, otherwise there's a resource leak). The behavior of Set assumes that it's safe to overwrite existing values, with no notification to the caller that a pre-existing value may have been discarded.

To safeguard against this today, (2) could issue a Get before a Set to check for the described race, but this results in additional locking overhead vs. if Bigcache provided a GetOrSet call in the spirit of sync.Map's LoadOrStore. Or, alternatively, perhaps a SetWithInfo call that includes the value that was overwritten. It's arguable that the locking overhead of Get-before-Set may be negligible if the cost to compute the value is high enough. Perhaps this is a case of YMMV. 

There are probably other use cases, this happens to be the one that I'm dealing with today.

**Comment by janisz** [comment]
I like this feature! @cristaloleg @mxplusb What do you think?

@jdef Would you like to provide a patch?

**Comment by mxplusb** [comment]
I like the idea of this feature, but I would prefer to not use a goroutine and that it be a blocking operation. The reason I think it should be a blocking operation because goroutines start to add complexity and I think the initial implementation should be simple and then it can be iterated upon once it's a bit more stable.

Aside from the goroutines, I am a bit concerned about the performance impact of this as a core API. This has the possibility to do a lot of branching which can really impact performance across the board, especially if it's locking for a long time.

That being said, those are things that come to mind and are in no way meant to discourage, just something to think about when working through it. I would be interested in seeing the API and seeing performance benchmarks on it as an initial implementation. 😄

**Comment by janisz** [comment]
> I like the idea of this feature, but I would prefer to not use a goroutine and that it be a blocking operation

Totally agree.

**Comment by cristaloleg** [comment]
Yep, sounds good, thank you @jdef.

Looks like everyone is happy with `GetOrSet(key string, newValue []byte) (oldValue []byte, err error)` API, which will be blocking.

From what I remember, we've `getWithoutLock` and `setWithoutLock`, and basically `GetOrSet` will be 'concatenation' of this methods (in some way). Any other objections? Should we touch `GetWithInfo` in some way?

**Comment by mxplusb** [comment]
> From what I remember, we've `getWithoutLock` and `setWithoutLock`, and basically `GetOrSet` will be 'concatenation' of this methods (in some way). Any other objections? Should we touch `GetWithInfo` in some way?

I personally think this should be a locking and blocking operation because it could introduce a race condition otherwise, which is where my comments about performance stemmed from. That being said, if it can be proved there's no race condition with a soak test (i.e. 10-100k concurrent `GetOrSet` calls on the same few objects) as part of the unit and acceptance tests then I don't think a locking operation should be necessary. I'll leave it up to the implementor to determine the right path on how to do that, though.

**Comment by cristaloleg** [comment]
@mxplusb no-no, don't get me wrong, `GetOrSet` will take a lock, I have been talking about implementation based on two useful functions :)

**Comment by mxplusb** [comment]
Ah no worries! I figure it'll be good to see who tackles it, how they approach it, and then go from there.
